Libplanet.Net/BlockCandidateTable.cs

```diff
@@ -18,79 +17,238 @@ namespace Libplanet.Net
+
+                            longestBranch ??= newBranch;
+                            longestBranch = CompareBranch(newBranch, longestBranch);
+                        }
+                    }
+                }
+
+                if (longestBranch is { })
+                {
+                    newBag.Add(longestBranch);
+                    BestBranch = longestBranch;
+                }
+
+                Branches = newBag;
+            }
         }
 
-        public bool TryRemove(BlockHeader header)
+        /// <summary>
+        /// Removes every <see cref="CandidateBranch{T}"/> in <see cref="Branches"/>.
+        /// </summary>
+        public void Clear()
         {
-            return _blocks.TryRemove(header, out _);
+            Branches = new ConcurrentBag<CandidateBranch<T>>();
+            BestBranch = null;
         }
 
-        public void Cleanup(Func<IBlockExcerpt, bool> predicate)
+        public bool Any() => Branches.Any();
+
+        /// <summary>
+        /// Check every blocks are continuous by <see cref="Block{T}.PreviousHash"/>-wise each
+        /// other.
+        /// <remarks>
+        /// This method does not check whether branch is appendable to <see cref="BlockChain{T}"/>.
+        /// </remarks>
+        /// </summary>
+        /// <param name="branch">A series of <see cref="Block{T}"/>.
+        /// </param>
+        /// <returns>
+        /// returns true if <paramref name="branch"/> is continuous or its count is 1; returns false
+        /// if <paramref name="branch"/> is not continuous or its count is 0.
+        /// </returns>
+        private static bool AreBlocksContinuous(IEnumerable<Block<T>> branch)
         {
-            foreach (var blockHeader in _blocks.Keys)
+            var enumerable = branch as Block<T>[] ?? branch.ToArray();
+
+            if (!enumerable.Any())
+            {
+                return false;
+            }
+
+            if (enumerable.Count() == 1)
             {
-                if (!predicate(blockHeader))
+                return true;
+            }
+
+            var precedingBlocks = enumerable.Skip(1);
+            foreach (var block in precedingBlocks)
+            {
+                if (!enumerable.Any(x => block.PreviousHash != null &&
+                                         x.Hash.Equals(block.PreviousHash.Value)))
                 {
-                    _blocks.TryRemove(blockHeader, out _);
+                    return false;
                 }
             }
+
+            return true;
         }
+
+        // FIXME: This method is duplicated with Swarm<T>.IsBlockNeeded()
+        private bool IsBlockNeeded(IBlockExcerpt block, IBlockExcerpt tip) =>
+            _canonicalChainComparer.Compare(block, tip) > 0;
+
+        private CandidateBranch<T> CompareBranch(
+            CandidateBranch<T> lf,
+            CandidateBranch<T> rf)
+            => _canonicalChainComparer.Compare(lf.Tip, rf.Tip) > 0 ? lf : rf;
```

Comment:
This is problematic. The logic is not argument order agnostic, i.e. `CompareBranch(a, b)` is not necessarily equal to `CompareBranch(b, a)`. This should be made clearer by using a different name for the method or via renaming of parameters. Think of a scenario where `CompareBranch(a, b)` might be equal to `0`.

Reply:
Yes, I've noticed where the `a == b` boundary exists while implementing `IsBlockNeeded()` and I somehow miss the same thing applied to `CompareBranch()`.😅

I'm not sure what would be the best name for this method. I've tried and commented on additional documentation.

https://github.com/planetarium/libplanet/blob/dedae5e3f869925a9d59afb9bfb1186369461197/Libplanet.Net/BlockCandidateTable.cs#L238-L259